Middle task tutorials (#58)
epiverse-trace/tutorials: episodes/create-forecast.Rmd

---
title: 'Create a short-term forecast'
teaching: 10
exercises: 2
---

```{r setup, echo= FALSE, warning = FALSE, message=FALSE}
require(EpiNow2)
require(ggplot2)
require(incidence2)
require(webshot)
webshot::install_phantomjs(force = TRUE)
options(mc.cores = 4)
```

:::::::::::::::::::::::::::::::::::::: questions 

- How do I create short-term forecasts of infections from case data?
- How do I account for incomplete reporting in forecasts?

::::::::::::::::::::::::::::::::::::::::::::::::

::::::::::::::::::::::::::::::::::::: objectives

- Learn how to make forecasts of cases using R package `EpiNow2`
- Learn how to include an observation process in estimation

::::::::::::::::::::::::::::::::::::::::::::::::

::::::::::::::::::::::::::::::::::::: prereq

## Prerequisites

+ Complete tutorial 'Quantifying transmissibility'

This tutorial has the following concept dependencies: 

**Statistics** : probability distributions, principle of Bayesian analysis. 

**Epidemic theory** : Effective reproduction number.

:::::::::::::::::::::::::::::::::

## Introduction

> Given case data, we can create estimates of the current and future number of cases by accounting for both delays in reporting and under reporting. In this lesson we will create short term forecasts of future using the inference framework in `{EpiNow2}`.

Reviewer: ```suggestion
Given case data, we can create estimates of the current and future number of cases by accounting for both delays in reporting and under reporting. In this lesson we will create short-term forecasts of future using the inference framework in `{EpiNow2}`.
```

I would add something on *how* a forecast is made. Perhaps something along the lines of:

"In order to make a statement about the future such as we do when we make a short-term forecast we need to make an assumption of how observations up to today are related to what we expect to happen in the future. The simplest way of doing so is to assume "no change", i.e. the reproduction number to remain the same in the future as last observed. Below we will make a short-term forecast by assuming the reproduction number will remain the same as it was estimated to be on the final date for which data was available."

As somewhat subtle point is that by doing so the reproduction already as considerable amount of uncertainty (as there is little to inform it on the last date with data). This behaviour can be controlled with the `future` argument in `rt_opts`, see https://github.com/epiforecasts/EpiNow2/blob/a3ff51da886b4c243b2b3d4e8f6011639968fd8f/R/create.R#L150 - the default is "latest"; using "estimate" will lead to less uncertain forecasts and "project" to more uncertain forecasts.

Perhaps this could/should be explained to user somehow?

Author: I've added a callout to explain the difference between "latest" and "estimate" 26608d89c207ae521a91537fa77a7dcb1083f9c1. How much detail do you want on the Rt model? Currently there isn't anything, the simplest option would be to just link to an external resource

Reviewer: Could maybe link to this in the package doc? https://epiforecasts.io/EpiNow2/reference/rt_opts.html
Or do you think that's too technical?

Author: I think it would be too technical based on the level the tutorials are currently written for,  but useful to highlight that the 'project' option exists?

Reviewer: Perhaps link to the "examples" vignette which contains an example on this? https://epiforecasts.io/EpiNow2/dev/articles/estimate_infections_options.html

> ## Create a nowcast and short-term forecast

Reviewer: ```suggestion
## Create short-term forecast
```

Either remove `nowcast` or explain/define. I don't think the term is used very consistently in the literature but to me it means correcting for right truncation in incomplete reporting data which can be done with the `estimate_truncation` function and is beyond the scope of this tutorial.

Author: Thanks for the clarification, I have removed `nowcast`

> The function `epinow()` described in the previous lesson is a wrapper for the function `estimate_infections()` used to estimate cases by date of infection. Using the same code in the previous lesson we can extract the now cast and short term-forecast using ...

Reviewer: ```suggestion
The function `epinow()` described in the previous lesson is a wrapper for the function `estimate_infections()` used to estimate cases by date of infection. Using the same code in the previous lesson we can extract the short term-forecast using ...
```

rt_log_sd <- convert_to_logsd(2, 1)

latent_period_fixed <- dist_spec(
  mean = 4, sd = 2,
  max = 20, distribution = "gamma"
)

log_mean <- convert_to_logmean(2, 1)
log_sd <- convert_to_logsd(2, 1)
reporting_delay_fixed <- dist_spec(
  mean = log_mean, sd = log_sd,
  max = 10, distribution = "lognormal"
)

generation_time_fixed <- dist_spec(
  mean = 3.6, sd = 3.1,
  max = 20, distribution = "lognormal"
)
```

```{r, message = FALSE, eval = TRUE}
reported_cases <- cases[1:90, ]
estimates <- epinow(
  reported_cases = reported_cases,
  generation_time = generation_time_opts(generation_time_fixed),
  delays = delay_opts(latent_period_fixed + reporting_delay_fixed),
  rt = rt_opts(prior = list(mean = rt_log_mean, sd = rt_log_sd))
)
```

We can visualise the estimates of the effective reproduction number and the estimated number of cases using `plot()`. The estimates are split into three categories:

+ **Estimate** : utilises all data,

+ **Estimate based on partial data** : estimates that are based less data are are therefore more uncertain,

+ **Forecast** : forecasts into the future. 

```{r}
plot(estimates)
```

### Incomplete observations 

> In the previous lesson we accounted for delays in reporting. In `EpiNow2` we also can account for incomplete observations ad in reality, 100% of cases are not reported.

Reviewer: ```suggestion
In the previous lesson we accounted for delays in reporting. In `EpiNow2` we also can account for incomplete observations as in reality, 100% of cases are not reported.
```

  delays = delay_opts(latent_period_fixed + reporting_delay_fixed),
  rt = rt_opts(prior = list(mean = rt_log_mean, sd = rt_log_sd)),
  obs = obs_opts(scale = obs_scale)
)
summary(estimates)
```

The estimates of transmissibility measures such as the effective reproduction number and rate of growth are similar (or the same in value) compared to when we didn't account for incomplete observations (see previous lesson). However the number of new confirmed cases by infection date has changed substantially in magnitude to reflect the assumption that only 40% of cases are in the data set.

We can also change the default distribution from Negative Binomial to Poisson, remove the default week effect and more. See `?EpiNow2::obs_opts` for more detail.
 

## Estimate hospitalisations and deaths

*Coming soon*

## Summary

Adding an observational process to account for incomplete reporting. There are range of different model options in `EpiNow2` that aren't covered in these lessons, see the [vignette](https://epiforecasts.io/EpiNow2/dev/articles/estimate_infections_options.html) for more details.

::::::::::::::::::::::::::::::::::::: keypoints 

> - Incomplete case reporting can be accounted for in estimates

Reviewer: maybe as first point something like:
- we can make short-term forecasts by making an assumption about the future behaviour of the reproduction number

)
```

```{r, message = FALSE, eval = TRUE}
reported_cases <- cases[1:90, ]
estimates <- epinow(
  reported_cases = reported_cases,
  generation_time = generation_time_opts(generation_time_fixed),
  delays = delay_opts(incubation_period_fixed + reporting_delay_fixed),
  rt = rt_opts(prior = list(mean = rt_log_mean, sd = rt_log_sd))
)
```

We can visualise the estimates of the effective reproduction number and the estimated number of cases using `plot()`. The estimates are split into three categories:

+ **Estimate** (green): utilises all data,

+ **Estimate based on partial data** (orange): estimates that are based less data are are therefore more uncertain,

+ **Forecast** (purple): forecasts into the future. 

```{r}
plot(estimates)
```

::::::::::::::::::::::::::::::::::::: callout
### Forecasting with estimates of $R_t$

By default, the short-term forecasts are created using the latest estimate of the reproduction number $R_t$. As this estimate is based on partial data, it has considerable uncertainty.  

The reproduction number can be changed to an estimate based on more data using `rt_ops()`:

```{r, eval = FALSE}
rt_opts(future = "estimate")
```

> The result will be less uncertain forecasts (as they are based on $R_t$ with a narrower uncertainty interval) but the forecasts will be based on less recent estimates of $R_t$.  

Reviewer: ```suggestion
The result will be less uncertain forecasts (as they are based on $R_t$ with a narrower uncertainty interval) but the forecasts will be based on less recent estimates of $R_t$ and assume no change sine then.
```

---
title: 'Create a short-term forecast'
teaching: 30
exercises: 30
editor_options: 
  chunk_output_type: console
---

```{r setup, echo = FALSE, warning = FALSE, message = FALSE}
library(EpiNow2)
withr::with_options(mc.cores = 4)
```

:::::::::::::::::::::::::::::::::::::: questions 

> - How do I create short-term forecasts of infections from case data?

Reviewer: ```suggestion
- How do I create short-term forecasts from case data?
```

```{r, echo = FALSE}
cases <- aggregate(
  cases_new ~ date,
  data = incidence2::covidregionaldataUK[, c("date", "cases_new")],
  FUN = sum
)
colnames(cases) <- c("date", "confirm")
rt_log_mean <- convert_to_logmean(2, 1)
rt_log_sd <- convert_to_logsd(2, 1)

incubation_period_fixed <- dist_spec(
  mean = 4, sd = 2,
  max = 20, distribution = "gamma"
)

log_mean <- convert_to_logmean(2, 1)
log_sd <- convert_to_logsd(2, 1)
reporting_delay_fixed <- dist_spec(
  mean = log_mean, sd = log_sd,
  max = 10, distribution = "lognormal"
)

generation_time_fixed <- dist_spec(
  mean = 3.6, sd = 3.1,
  max = 20, distribution = "lognormal"
)
```

```{r, message = FALSE, eval = TRUE}
reported_cases <- cases[1:90, ]
estimates <- epinow(
  reported_cases = reported_cases,
  generation_time = generation_time_opts(generation_time_fixed),
  delays = delay_opts(incubation_period_fixed + reporting_delay_fixed),
  rt = rt_opts(prior = list(mean = rt_log_mean, sd = rt_log_sd))
)
```

We can visualise the estimates of the effective reproduction number and the estimated number of cases using `plot()`. The estimates are split into three categories:

+ **Estimate** (green): utilises all data,

> + **Estimate based on partial data** (orange): estimates that are based less data are are therefore more uncertain,

Reviewer: ```suggestion
+ **Estimate based on partial data** (orange): estimates that are based less data are therefore more uncertain,
```

+ **Forecast** (purple): forecasts into the future. 

```{r}
plot(estimates)
```

::::::::::::::::::::::::::::::::::::: callout
### Forecasting with estimates of $R_t$

By default, the short-term forecasts are created using the latest estimate of the reproduction number $R_t$. As this estimate is based on partial data, it has considerable uncertainty.  

> The reproduction number can be changed to an estimate based on more data using `rt_ops()`:

Reviewer: ```suggestion
The reproduction number that is projected into the future can be changed to a less recent estimate based on more data using `rt_ops()`:
```

> The reproduction number that is projected into the future can be changed to a less recent estimate based on more data using `rt_ops()`:

Reviewer 2: ```suggestion
The reproduction number that is projected into the future can be changed to a less recent estimate based on more data using `rt_opts()`:
```

```{r, eval = FALSE}
rt_opts(future = "estimate")
```

The result will be less uncertain forecasts (as they are based on $R_t$ with a narrower uncertainty interval) but the forecasts will be based on less recent estimates of $R_t$ and assume no change since then.

> Additionally, there is the option to project the value of $R_t$ into the future using a model by setting `future = "project"`. As this option uses a model to forecast the value of $R_t$, the result will be forecasts that are more uncertain than `estimate`, for an example [see here](https://epiforecasts.io/EpiNow2/dev/articles/estimate_infections_options.html#projecting-the-reproduction-number-with-the-gaussian-process).

Reviewer: ```suggestion
Additionally, there is the option to project the value of $R_t$ into the future using a generic model by setting `future = "project"`. As this option uses a model to forecast the value of $R_t$, the result will be forecasts that are more uncertain than `estimate`, for an example [see here](https://epiforecasts.io/EpiNow2/dev/articles/estimate_infections_options.html#projecting-the-reproduction-number-with-the-gaussian-process).
```

::::::::::::::::::::::::::::::::::::::::::::::::

### Incomplete observations 

In the previous tutorial we accounted for delays in reporting. In `EpiNow2` we also can account for incomplete observations as in reality, 100% of cases are not reported.

We will pass another input into `epinow()` called `obs` to define an observation model. The format of `obs` must be the `obs_opt()` function (see `?EpiNow2::obs_opts` for more detail). 

> We believe the COVID-19 outbreak data from the previous tutorial do not include all reported cases. We believe that we only observe 40% of cases. To specify this in the observation model, we must pass a scaling factor with a mean and standard deviation. If we assume that 40% of cases are in the case data (with standard deviation 1%), then we specify the `scale` input to `obs_opts()` as follows:

Reviewer: ```suggestion
Let's say we believe the COVID-19 outbreak data from the previous tutorial do not include all reported cases. We believe that we only observe 40% of cases. To specify this in the observation model, we must pass a scaling factor with a mean and standard deviation. If we assume that 40% of cases are in the case data (with standard deviation 1%), then we specify the `scale` input to `obs_opts()` as follows:
```

ebola_incubation_period <- dist_spec(
  mean = 2.487, sd = 0.330,
  mean_sd = 0.5, sd_sd = 0.5,
  max = 20, distribution = "lognormal"
)

ebola_generation_time <- dist_spec(
  mean = 15.3, sd = 10.1,
  mean_sd = 0.5, sd_sd = 0.5,
  max = 30, distribution = "gamma"
)
```

As we want to also create a two week forecast, we specify `horizon = 14` to forecast 14 days instead of the default 7 days. 

```{r, eval=FALSE,echo=FALSE}
# read data from the tutorial repository R project
ebola_confirmed <-
  read_csv(file.path("episodes", "data", "ebola_cases.csv"))
```

```{r,eval=FALSE,echo=TRUE, message = FALSE}
# read data
# e.g.: if path to file is data/raw-data/ebola_cases.csv then:
ebola_confirmed <-
  read_csv(here::here("data", "raw-data", "ebola_cases.csv"))

# format date column
ebola_cases$date <- as.Date(ebola_cases$date)

ebola_estimates <- epinow(
  reported_cases = ebola_cases[1:120, ], # first 3 months of data only
  generation_time = generation_time_opts(ebola_generation_time),
  delays = delay_opts(ebola_incubation_period),
  # horizon needs to be 14 days to create two week forecast (default is 7 days)
  horizon = 14
)

summary(ebola_estimates)
```

> The effective reproduction number $R_t$ estimate (on the last date of the data) is 1.7 (1 - 3). The exponential growth rate of case numbers is 0.042 (0.0019 - 0.094).

Reviewer: Should we not create this by running the code above (which would also serve to check that it works as written)?

Reviewer 2: indeed, recent commits introduced issues given the absence of this running. I reactivated this in https://github.com/epiverse-trace/tutorials/pull/58/commits/8cf8d31895bd30043cc078d49b584ad704e17b57 let's wait for the workflow to pass.

Reviewer 2: We are getting this error in the workflow:

```
Quitting from lines 324-327 (/home/runner/work/tutorials/tutorials/episodes/create-forecast.Rmd)
Error: 
! in callr subprocess.
Caused by error in `eval(parse_only(code), envir = envir)`:
! object 'ebola_estimates' not found
```

Although the previous chunk and in-line code in line 324 runs ok locally. See code in context:

https://github.com/epiverse-trace/tutorials/blob/8cf8d31895bd30043cc078d49b584ad704e17b57/episodes/create-forecast.Rmd#L295-L327

I'm building this locally if any other issue appears. The goal is to get the `ebola_estimates` output.

Reviewer 2: is possible that line 299 may not be reading input data as we expect. Reference workbench repo reads data from a path starting at "data" and not at "episodes" folder ([website](https://www.michaelc-m.com/Rewrite-R-ecology-lesson/04-putting-it-together.html#reading-the-new-surveys-data), [path](https://github.com/MCMaurer/Rewrite-R-ecology-lesson/tree/96728131fe86f5a2913b4040169192cd9fa7eed2/episodes/data/raw))

Reviewer: I think that's spot on - should be `file.path("data", "ebola_cases.csv")`

Reviewer 2: tested in a [minimal repo](https://github.com/avallecam/workbench-spanish/pull/3/files#diff-ddc674eebc510c36c9648ce8c54a64cfbbb92cc8f2d630ced5dce13f0d8901a7R27) and this seems to work.

Reviewer 2: applied in https://github.com/epiverse-trace/tutorials/pull/58/commits/c0aa6d2914de3d70a77d88a68c9f1d375b4bb73b waiting for workflow response.

Reviewer 2: success! I'll resolve this conversation

ebola_incubation_period <- dist_spec(
  mean = 2.487, sd = 0.330,
  mean_sd = 0.5, sd_sd = 0.5,
  max = 20, distribution = "lognormal"
)

ebola_generation_time <- dist_spec(
  mean = 15.3, sd = 10.1,
  mean_sd = 0.5, sd_sd = 0.5,
  max = 30, distribution = "gamma"
)
```

As we want to also create a two week forecast, we specify `horizon = 14` to forecast 14 days instead of the default 7 days. 

```{r, eval=TRUE,echo=FALSE}
# read data from the tutorial repository R project
ebola_cases <-
  read.csv(file.path("data", "ebola_cases.csv"))
```

```{r,eval=FALSE,echo=TRUE}
# read data
# e.g.: if path to file is data/raw-data/ebola_cases.csv then:
ebola_cases <-
  read.csv(here::here("data", "raw-data", "ebola_cases.csv"))
```

```{r,eval=TRUE,echo=TRUE, message = FALSE}
# format date column
ebola_cases$date <- as.Date(ebola_cases$date)

ebola_estimates <- epinow(
  reported_cases = ebola_cases[1:120, ], # first 3 months of data only
  generation_time = generation_time_opts(ebola_generation_time),
  delays = delay_opts(ebola_incubation_period),
  # horizon needs to be 14 days to create two week forecast (default is 7 days)
  horizon = 14
)

summary(ebola_estimates)
```

> The effective reproduction number $R_t$ estimate (on the last date of the data) is `r summary(ebola_estimates)$estimate[summary(ebola_estimates)$measure=="Effective reproduction no."]`. The exponential growth rate of case numbers is `r summary(ebola_estimates)$estimate[summary(ebola_estimates)$measure=="Rate of growth"]`.

Reviewer: These are `data.table`s so could shorten this to 
```suggestion
The effective reproduction number $R_t$ estimate (on the last date of the data) is `r summary(ebola_estimates)[measure=="Effective reproduction no."]$estimate`. The exponential growth rate of case numbers is `r summary(ebola_estimates)[measure=="Rate of growth"]$estimate`.
```

Reviewer 2: yes, it works!
